reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -319,492 +244,554 @@ public String toString() {
+				if (candidate.tryEvict(evictionTimestamp, parent.ttlMillis)) {
+					idleQueue.remove(candidate);
+					decrementAndGet();
+				}
 			}
 		}
 
-		@Override
-		public void dispose() {
-			if (exec != null) {
-				if (this != SHUTDOWN && !parent.shutdown) {
-					//in case of work, re-create an ActiveWorker
-					DeferredFacade deferredFacade = parent.deferredFacades.poll();
-					if (deferredFacade != null) {
-						deferredFacade.setService(this);
+		/**
+		 * Pick a {@link BoundedState}, prioritizing idle ones then spinning up a new one if enough capacity.
+		 * Otherwise, picks an active one by taking from a {@link PriorityQueue}. The picking is
+		 * optimistically re-attempted if the picked slot cannot be marked as picked.
+		 *
+		 * @return the picked {@link BoundedState}
+		 */
+		BoundedState pick() {
+			for (;;) {
+				int a = get();
+				if (a == DISPOSED) {
+					return CREATING; //synonym for shutdown, since the underlying executor is shut down
+				}
+
+				if (!idleQueue.isEmpty()) {
+					//try to find an idle resource
+					BoundedState bs = idleQueue.pollLast();
+					if (bs != null && bs.markPicked()) {
+						busyQueue.add(bs);
+						return bs;
 					}
-					else {
-						//if no more work, the service is put back at end of the cached queue and new expiry is started
-						CachedServiceExpiry e = new CachedServiceExpiry(this,
-								System.currentTimeMillis() + parent.ttlSeconds * 1000L);
-						parent.idleServicesWithExpiry.offerLast(e);
-						if (parent.shutdown) {
-							if (parent.idleServicesWithExpiry.remove(e)) {
-								exec.shutdownNow();
-							}
-						}
+					//else optimistically retry
+				}
+				else if (a < parent.maxThreads) {
+					//try to build a new resource
+					if (compareAndSet(a, a + 1)) {
+						ScheduledExecutorService s = Schedulers.decorateExecutorService(parent, parent.createBoundedExecutorService());
+						BoundedState newState = new BoundedState(this, s);
+						newState.markPicked(); //no need to check return value as this one is brand new
+						busyQueue.add(newState);
+						return newState;
+					}
+					//else optimistically retry
+				}
+				else {
+					//pick the least busy one
+					BoundedState s;
+					s = busyQueue.poll();
+					if (s != null && s.markPicked()) {
+						busyQueue.add(s); //put it back in the queue with updated priority
+						return s;
 					}
+					//else optimistically retry
 				}
 			}
 		}
 
-		@Override
-		public Object scanUnsafe(Attr key) {
-			if (key == Attr.NAME) return parent.scanUnsafe(key);
-			if (key == Attr.PARENT) return parent;
-			if (key == Attr.TERMINATED || key == Attr.CANCELLED) return isDisposed();
-			if (key == Attr.CAPACITY) {
-				//assume 1 if unknown, otherwise use the one from underlying executor
-				Integer capacity = (Integer) Schedulers.scanExecutor(exec, key);
-				if (capacity == null || capacity == -1) return 1;
-			}
-			return Schedulers.scanExecutor(exec, key);
+		void setIdle(BoundedState boundedState) {
```

Comment:
We need to be careful with the order here, otherwise:
1. we add it to `idleQueue`
2. in another thread, `pick()` polls it and adds to the busy queue and returns
3. next statement removes it from the busy queue

Reply:
indeed. reversing the operations should be safer: remove from `busyQueue` (it is now nowhere to be seen by a concurrent process) -> add to `idleQueue` 👍

Reply:
(I probably should add a test for that)